Add global Rakuyomi entry point (toolbar / Dispatcher friendly)

cramoisi opened the pull request:
### **User description**
This PR introduces a single, unified global entry point for Rakuyomi, making it possible to open the plugin from anywhere in KOReader (reader, file manager, gestures, toolbar plugins such as Project Title).

**What’s included**

- Persistent Rakuyomi instance
- - The current plugin instance is stored (Rakuyomi.instance) during init().
- - This allows Rakuyomi to be safely invoked from external contexts.

- New unified entry point: Rakuyomi.openFromToolbar()
- - Handles both ReaderUI and non-ReaderUI contexts.
- - Closes the reader when needed and returns to the Rakuyomi library view.
- - Includes safety checks (missing instance / backend not initialized).

- Redirection of the existing action
- - onStartLibraryView() now delegates to openFromToolbar().
- - This preserves the existing start_library_view action while unifying the logic behind it.

- Dispatcher / toolbar compatibility
- - Rakuyomi can now be reliably bound to:
- - - gestures,
- - - QuickMenu,
- - - toolbar plugins (e.g. Project Title),
- - without duplicating logic or relying on menu-only entry points.

**Why this change**

Previously, Rakuyomi was primarily menu-driven and difficult to integrate cleanly with global UI elements.
This PR makes Rakuyomi behave like a first-class, globally accessible app in KOReader, without breaking existing behavior.

**Compatibility**

- No existing APIs are removed or renamed.
- Existing menu entries and actions continue to work as before.
- Changes are self-contained and low-risk.


___

### **PR Type**
Enhancement


___

### **Description**
- Introduces persistent global Rakuyomi instance for unified entry point

- Adds `openFromToolbar()` static method for dispatcher/toolbar compatibility

- Refactors `onStartLibraryView()` to delegate to new entry point

- Handles both ReaderUI and non-ReaderUI contexts with proper cleanup

- Exports Rakuyomi module for external access via `package.loaded`


___

### Diagram Walkthrough


```mermaid
flowchart LR
  A["Rakuyomi.instance<br/>stored in init()"] -- "delegates to" --> B["openFromToolbar()"]
  B -- "checks backend" --> C["Error Dialog"]
  B -- "if ReaderUI" --> D["Close Reader"]
  B -- "show library" --> E["LibraryView"]
  D -- "callback" --> E
  F["Dispatcher/Toolbar<br/>Gestures"] -- "calls" --> B
```



<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>main.lua</strong><dd><code>Implement global entry point with persistent instance</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

frontend/rakuyomi.koplugin/main.lua

<ul><li>Added <code>InfoMessage</code> import for error notifications<br> <li> Introduced <code>Rakuyomi.instance</code> class variable to store persistent plugin <br>instance<br> <li> Modified <code>init()</code> to store self reference in <code>Rakuyomi.instance</code><br> <li> Refactored <code>onStartLibraryView()</code> to call new <code>openFromToolbar()</code> method<br> <li> Created new static method <code>openFromToolbar()</code> with safety checks and <br>context-aware handling<br> <li> Added module export via <code>package.loaded["rakuyomi"]</code> for external access</ul>


</details>


  </td>
  <td><a href="https://github.com/tachibana-shin/rakuyomi/pull/16/files#diff-05804ef67088286fb070b9193ea74036d93c03c0e2c5317d10c6b9b73867b88d">+39/-13</a>&nbsp; </td>

</tr>
</table></td></tr></tbody></table>

</details>

___

cramoisi committed:
Update main.lua

cramoisi committed:
Update main.lua

cramoisi commented on frontend/rakuyomi.koplugin/main.lua:
```diff
@@ -87,4 +83,34 @@ function Rakuyomi:openLibraryView()
+        UIManager:show(InfoMessage:new{
+            text = "Rakuyomi: no instance",
+            timeout = 2,
+        })
```
I was using it for debug purpose. It doesn't need to stay. I removed it.

tachibana-shin commented on frontend/rakuyomi.koplugin/main.lua:
```diff
@@ -21,11 +22,16 @@ local Rakuyomi = InputContainer:extend({
 -- so we should register to the menu accordingly
 function Rakuyomi:init()
+
+  Rakuyomi.instance = self
```
remove newline before this

tachibana-shin commented on frontend/rakuyomi.koplugin/main.lua:
```diff
@@ -37,27 +43,17 @@ function Rakuyomi:init()
+    Rakuyomi.openFromToolbar()
 end
 
+
```
lint

tachibana-shin commented on frontend/rakuyomi.koplugin/main.lua:
```diff
@@ -87,4 +83,29 @@ function Rakuyomi:openLibraryView()
+    end
+
+    if self.ui and self.ui.name == "ReaderUI" then
+        MangaReader:closeReaderUi(function()
```
initializeFromReaderUI needs to be called it helps rakuyomi register events

cramoisi committed:
Update main.lua

I’ve added a small guard so openFromToolbar() ensures MangaReader:initializeFromReaderUI(self.ui) has been called when running under ReaderUI before closing the reader. This guarantees Rakuyomi’s ReaderUI hooks/events are registered even when invoked via a global entry point (Project Title / gestures).

cramoisi committed:
Update main.lua

modif suggested by coderabbitai[bot]

cramoisi commented on frontend/rakuyomi.koplugin/main.lua:
```diff
@@ -87,4 +83,35 @@ function Rakuyomi:openLibraryView()
+    local self = Rakuyomi.instance
+    if not self then
+        return
+    end
```
done

cramoisi committed:
Update main.lua

Added a small re-entrancy guard to openFromToolbar() to prevent multiple rapid invocations (toolbar / gesture). Also ensured initializeFromReaderUI() is always called when invoked from ReaderUI before closing the reader, as suggested by bot

cramoisi committed:
Update main.lua

cramoisi committed:
Refactor Rakuyomi toolbar opening logic

cramoisi committed:
formatting & deduplicate initializeFromReaderUI

Add patching mechanism for ReaderUI initialization

cramoisi committed:
Merge branch 'main' into patch-1

tachibana-shin commented:
something is breaking this pr on koreader 11.2025 i compared the logic but it doesn't seem to change much
